### PR TITLE
bitwarden: Master Password Reprompt + Cleanup + Refactoring

### PR DESCRIPTION
**Pull request by eth-p**
## Description

These four commits implement the "Master password re-prompt" option for Raycast's Bitwarden extension.

The first commit is a refactoring of how Bitwarden vault sessions are handled, and aims to decouple the session management from the Search interface while also providing a more React-like usage. This was needed to prevent prop-drilling and make it easier to implement the master password re-prompt feature.  

The second commit cleans up the clipboard password copy code from my last pull request. I forgot to push it 😅

The third and forth commits implement the re-prompt feature and a setting for temporarily disabling re-prompt dialogs for a specified period of time after entering the master password.

*When reviewing, I would recommend looking at it commit-by-commit.*
It was structured so each commit is a single feature :)

## Checklist

- [x] I read the [extension guidelines](https://developers.raycast.com/basics/prepare-an-extension-for-store)
- [x] I read the [documentation about publishing](https://developers.raycast.com/basics/publish-an-extension)
- [x] I ran `npm run build` and [tested this distribution build in Raycast](https://developers.raycast.com/basics/prepare-an-extension-for-store#metadata-and-configuration)
- [x] I checked that files in the `assets` folder are used by the extension itself
- [x] I checked that assets used by the `README` are placed outside of the `metadata` folder

### CODE REVIEW

File: extensions/bitwarden/src/session.tsx
Comment (pomdtr) [inline extensions/bitwarden/src/session.tsx:116]: The `bw status` command is buggy in the cli, we cannot rely on it. The best approach I found out was to run the command (ex `bw list items`), and to show the form if an error occurs.

See https://github.com/bitwarden/clients/issues/2729 for more infos.

**Review by pomdtr — CHANGES_REQUESTED** [review]
Thanks for for your work, all of these changes are welcome !

I only highlighted a single issue, please take a look at it.

File: extensions/bitwarden/src/session.tsx
Comment (jomifepe) [inline extensions/bitwarden/src/session.tsx:105]: I'm a bit confused with this approach. All these functions are being passed as the `useState`'s initial state, which means that they are not reactive to the component's scope, and also, they don't need to be changed programmatically, so no need for them to be here. I'm not sure if there's anything inside the functions that can change at runtime, but nonetheless, these should be moved outside the useState initializer, and passed to the provider below, alongside the rest of the values (e.g. token, isLoading, etc). Same for the `api` variable, which is already passed as prop.

**Review by jomifepe — COMMENTED** [review]
Great idea on the reprompt feature 🚀  

Although, I pointed out some concerns about the code.

File: extensions/bitwarden/src/session.tsx
Comment (eth-p) [reply extensions/bitwarden/src/session.tsx:116]: Thanks for pointing that out! I'll change it and add a comment to capture the context.

File: extensions/bitwarden/src/actions.tsx
Comment (eth-p) [reply extensions/bitwarden/src/actions.tsx:71]: Good catch on that! It was supposed to be more of a wrapper function than a hook, but the `useNavigation` within the `reprompt` function does make it a bit problematic. I'll take your suggestion on this and turn it into a hook instead.

File: extensions/bitwarden/src/session.tsx
Comment (eth-p) [reply extensions/bitwarden/src/session.tsx]: I wanted to make it a bit less verbose by just keeping everything within a single object and updating that, rather than having both a `Session` object for public API consumption and a separate `SessionState` object for holding the state data itself.

In hindsight, it was a pretty bad approach that made far less sense for the sake of brevity. I split them back up into a `Session` class and `SessionState` interface in my latest round of changes, and wired it up appropriately so the `Session` exposes only the necessary parts of the `SessionState`.

File: extensions/bitwarden/src/actions.tsx
Comment (eth-p) [reply extensions/bitwarden/src/actions.tsx]: Changed it to be a hook instead, thanks for the suggestion!

File: extensions/bitwarden/src/session.tsx
Comment (eth-p) [reply extensions/bitwarden/src/session.tsx]: Context is now captured:

https://github.com/raycast/extensions/pull/3413/files#diff-59101489bd951ec97dd45c500bb43bc2775e174d64b127b034e16df3f41ec595R121-R145

**Comment by pernielsentikaer** [comment]
I lost track of the status. Does anyone need to recheck it? Please ping if anyone needs to do anything 😊

**Comment by pomdtr** [comment]
Please do not merge it yet, I want to take another look at it :)

**Comment by pernielsentikaer** [comment]
Sure, I'll await you 😊

File: extensions/bitwarden/src/actions.tsx
Comment (jomifepe) [reply extensions/bitwarden/src/actions.tsx]: Looks good 👍

File: extensions/bitwarden/src/session.tsx
Comment (jomifepe) [reply extensions/bitwarden/src/session.tsx]: Cool, nice work 👍 

I don't really like the class approach, this is React after all, and that could easily be handled by using hooks and some functions, but I'll leave that to be evaluated by @pomdtr.

**Comment by pomdtr** [comment]
I'll revesit this PR tomorrow, sorry for the delay @eth-p

**Comment by pernielsentikaer** [comment]
@pomdtr did you have time to revisit this yet?

**Comment by pernielsentikaer** [comment]
Let's check with @pomdtr one last time since he said he wanted to test it out 😊

**Comment by pomdtr** [comment]
Hey, I finally took the time time to test it (sorry for the delay, I will probably transfer the ownership of the extension soon).

It looks good to me ! However I noticed a bug after messing up the initial setup. Could you please fix the label of the toast ?

image

**Comment by jomifepe** [comment]
> Hey, I finally took the time time to test it (sorry for the delay, I will probably transfer the ownership of the extension soon, @jomifepe would you be interested ?).
> 
> It looks good to me ! However I noticed a bug after messing up the initial setup. Could you please fix the label of the toast ?
> 
> image

Hey, sorry for the delay. Do you mean transfer ownership for GitHub issues and PR's, not on the Raycast store right?

**Comment by pomdtr** [comment]
No problem. If someone steps up for handling issues and prs, he should probably steps up as the author of the extension. I'm fine beeing relegated to a contributor :)

**Comment by pernielsentikaer** [comment]
Should we agree on transferring the extension to @jomifepe and set @pomdtr as a contributor?

**Comment by pomdtr** [comment]
If @jomifepe agrees to takes ownership, go ahead. If not, has we've discussed on slack, @pernielsentikaer could also be the maintainer.

**Comment by pernielsentikaer** [comment]
Voting for @jomifepe too!

**Comment by jomifepe** [comment]
@pernielsentikaer @pomdtr Yeah I can take the ownership 🙂

**Review by pernielsentikaer — APPROVED** [review]
Let's make the change 🫣

**Pull request merged**
